File: internal/content/frequency.go
package content

import (
	"bufio"
	"fmt"
	"io/fs"
	"path"
	"strconv"
	"strings"

	"github.com/sebastiancaraballo/polyglot/internal/model"
)

// loadFrequency reads a target-language word-frequency list from
// content/<lang>/frequency.tsv (rank<TAB>word<TAB>reading<TAB>count), skipping
// blank lines and '#' comments. The list is parsed and validated but is not yet
// wired into card scheduling — that is the frequency-backfill follow-up.
func loadFrequency(fsys fs.FS, lang string) ([]model.FreqEntry, error) {
	file := path.Join("content", lang, "frequency.tsv")
	data, err := fs.ReadFile(fsys, file)
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", file, err)
	}

	var entries []model.FreqEntry
	sc := bufio.NewScanner(strings.NewReader(string(data)))
	sc.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	line := 0
	for sc.Scan() {
		line++
		text := sc.Text()
		if text == "" || strings.HasPrefix(text, "#") {
			continue
		}
		fields := strings.Split(text, "\t")
		if len(fields) != 4 {
			return nil, fmt.Errorf("%s:%d: expected 4 tab-separated fields, got %d", file, line, len(fields))
		}
		rank, err := strconv.Atoi(fields[0])
		if err != nil {
			return nil, fmt.Errorf("%s:%d: invalid rank %q: %w", file, line, fields[0], err)
		}
		count, err := strconv.Atoi(fields[3])
		if err != nil {
			return nil, fmt.Errorf("%s:%d: invalid count %q: %w", file, line, fields[3], err)
		}
		entries = append(entries, model.FreqEntry{
			Rank:    rank,
			Word:    fields[1],
			Reading: fields[2],
			Count:   count,
		})
	}
	if err := sc.Err(); err != nil {
		return nil, fmt.Errorf("read %s: %w", file, err)
	}
	return entries, nil
}

File: core/src/content/frequency.rs
use super::fsys::ContentFs;
use super::LoadError;
use crate::model::FreqEntry;

/// Reads a target-language word-frequency list from `<lang>/frequency.tsv`.
pub(super) fn load_frequency(
    fsys: &dyn ContentFs,
    lang: &str,
) -> Result<Vec<FreqEntry>, LoadError> {
    let file = format!("{lang}/frequency.tsv");
    let data = fsys
        .read(&file)
        .map_err(|e| LoadError::new(format!("read {file}: {e}")))?;
    parse_frequency(&file, &data)
}

/// Parses a frequency list (`rank<TAB>word<TAB>reading<TAB>count`), skipping
/// blank lines and `#` comments.
pub(super) fn parse_frequency(file: &str, data: &[u8]) -> Result<Vec<FreqEntry>, LoadError> {
    let text = std::str::from_utf8(data)
        .map_err(|e| LoadError::new(format!("read {file}: invalid utf-8: {e}")))?;

    let mut entries = Vec::new();
    for (i, line) in text.lines().enumerate() {
        let lineno = i + 1;
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let fields: Vec<&str> = line.split('\t').collect();
        if fields.len() != 4 {
            return Err(LoadError::new(format!(
                "{file}:{lineno}: expected 4 tab-separated fields, got {}",
                fields.len()
            )));
        }
        let rank: i64 = fields[0].parse().map_err(|e| {
            LoadError::new(format!(
                "{file}:{lineno}: invalid rank {:?}: {e}",
                fields[0]
            ))
        })?;
        let count: i64 = fields[3].parse().map_err(|e| {
            LoadError::new(format!(
                "{file}:{lineno}: invalid count {:?}: {e}",
                fields[3]
            ))
        })?;
        entries.push(FreqEntry {
            rank,
            word: fields[1].to_string(),
            reading: fields[2].to_string(),
            count,
        });
    }
    Ok(entries)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_and_skips_comments_and_blanks() {
        let tsv = "# comment\n1\t\u{79c1}\twatashi\t100\n\n2\t\u{306e}\tno\t80\n";
        let entries = parse_frequency("f.tsv", tsv.as_bytes()).unwrap();
        assert_eq!(entries.len(), 2);
        assert_eq!(entries[0].rank, 1);
        assert_eq!(entries[0].reading, "watashi");
        assert_eq!(entries[1].count, 80);
    }

    #[test]
    fn rejects_wrong_field_count() {
        let tsv = "1\tword\tonly-three\n";
        assert!(parse_frequency("f.tsv", tsv.as_bytes()).is_err());
    }
}
